### PR TITLE
TR-2961 text reader asset extraction

### PR DESCRIPTION
**Pull request by poyuki**
# [TR-2961](https://oat-sa.atlassian.net/browse/TR-2961)

## Summary
Need to create functionality which allow publish inline base64 encoded PCI Text Reader item assets to cdn
 
## How to test
* update extension by branch
* create `item` with `Text Reader` with image
* create `test` with created in prev step `item`
* publish this `test` as local
* run published delivery and check in browser inspector image `src`

For affecting local `env` without `cdn` need to update `taoQtiItem/model/compile/QtiAssetReplacer/NullQtiItemAssetReplacer.php`
by next code 
```
class NullQtiItemAssetReplacer extends ConfigurableService implements QtiItemAssetReplacer
{
    public function shouldBeReplaced(PackedAsset $packetAsset): bool
    {
        return $packetAsset->getMediaAsset()->getMediaSource() instanceof InlineSource;
    }

    public function replace(PackedAsset $packetAsset, $itemId): PackedAsset
    {
        $mediaAsset = $packetAsset->getMediaAsset();
        $filename = $mediaAsset->getMediaSource()->getBaseName($mediaAsset->getMediaIdentifier());

        return $packetAsset->setReplacedBy($filename);
    }
}
```

### CODE REVIEW

File: model/qti/CustomInteractionAsset/DataUrlMimeTypeDecoder.php [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+     * @param string $dataUrl
+     * @return string|null
+     * @throws NotDataUrlProvidedException
+     */
+    public static function decodeToAssetType(string $dataUrl): ?string
+    {
+        if (self::checkIsDataUrl($dataUrl)) {
+            $mimeType = mime_content_type($dataUrl);
+
+            return self::mapMimeTypeToAsset($mimeType);
```
Comment (wazelin) [inline model/qti/CustomInteractionAsset/DataUrlMimeTypeDecoder.php:94]: `$mimeType` can potentially be a `false`.

Comment (poyuki) [reply model/qti/CustomInteractionAsset/DataUrlMimeTypeDecoder.php]: fixed

File: model/qti/CustomInteractionAsset/ExtendedCustomInteractionAssetExtractorAllocator.php [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+use oat\taoQtiItem\model\qti\CustomInteractionAsset\Extractor\BaseExtendedCustomInteractionAssetExtractor;
+use oat\taoQtiItem\model\qti\CustomInteractionAsset\Extractor\TextReaderExtendedAssetExtractor;
+use oat\taoQtiItem\model\qti\interaction\CustomInteraction;
+use oat\taoQtiItem\model\qti\interaction\ImsPortableCustomInteraction;
+use oat\taoQtiItem\model\qti\interaction\PortableCustomInteraction;
+
+/**
+ * @author Kiryl Poyu <kyril.poyu@taotesting.com>
+ */
+class ExtendedCustomInteractionAssetExtractorAllocator
```
Comment (wazelin) [inline model/qti/CustomInteractionAsset/ExtendedCustomInteractionAssetExtractorAllocator.php:36]: This looks more like an abstract factory to me, WDYT?

I mean, it clearly _creates_ an extractor service.

Comment (poyuki) [reply model/qti/CustomInteractionAsset/ExtendedCustomInteractionAssetExtractorAllocator.php:36]: I do not think so. In general, this class is used just to determine which implementation of the base class we need. An abstract factory can be used as part of this flow if we have a complex process for creating descendants.

File: model/qti/CustomInteractionAsset/ExtendedCustomInteractionAssetExtractorAllocator.php [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+        if (
+            $interaction instanceof ImsPortableCustomInteraction
+            || $interaction instanceof PortableCustomInteraction
+        ) {
+            if ($interaction->getTypeIdentifier() === self::TEXT_READER_INTERACTION) {
+                return new TextReaderExtendedAssetExtractor($interaction);
+            }
+        }
+
+        return null;
```
Comment (wazelin) [inline model/qti/CustomInteractionAsset/ExtendedCustomInteractionAssetExtractorAllocator.php:52]: Instead of a `null`. WDYT of returning a no-op implementation of the interface?

Comment (poyuki) [reply model/qti/CustomInteractionAsset/ExtendedCustomInteractionAssetExtractorAllocator.php:52]: probably we can, but i don't see any real profit from this. We just remove 1 `if` but create empty class which should return empty array for this iteration (IMO it looks strange)
```php
if ($assetExtractor !== null) {
            $extractedAssets = $assetExtractor->extract();
            foreach ($extractedAssets as $asset) {
                $this->addAsset('pci-resource', $asset);
            }
        }
```

Comment (wazelin) [reply model/qti/CustomInteractionAsset/ExtendedCustomInteractionAssetExtractorAllocator.php:52]: That's the point of a no-op service. Yes, you have one `if` now, but you may as well get some other at some other place.
null-objects instead of `null` values allow reducing the complexity of the code.

I won't be pushing for that change any more, but suggest you think about it a bit :)

File: model/qti/CustomInteractionAsset/ExtendedCustomInteractionAssetExtractorAllocator.php [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+/**
+ * @author Kiryl Poyu <kyril.poyu@taotesting.com>
+ */
+class ExtendedCustomInteractionAssetExtractorAllocator
+{
+    public const TEXT_READER_INTERACTION = 'textReaderInteraction';
+
+    public static function allocateExtractor(
+        CustomInteraction $interaction
+    ): ?BaseExtendedCustomInteractionAssetExtractor {
```
Comment (wazelin) [inline model/qti/CustomInteractionAsset/ExtendedCustomInteractionAssetExtractorAllocator.php:42]: I'd prefer if this method returned an interface.
An abstract class is good to have a common piece of logic. A return type should first of define a public contract, which is the main purpose of the interfaces.

File: model/qti/CustomInteractionAsset/Extractor/BaseExtendedCustomInteractionAssetExtractor.php [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+ */
+
+namespace oat\taoQtiItem\model\qti\CustomInteractionAsset\Extractor;
+
+use oat\taoQtiItem\model\qti\interaction\ImsPortableCustomInteraction;
+use oat\taoQtiItem\model\qti\interaction\PortableCustomInteraction;
+
+/**
+ * @author Kiryl Poyu <kyril.poyu@taotesting.com>
+ */
```
Comment (wazelin) [inline model/qti/CustomInteractionAsset/Extractor/BaseExtendedCustomInteractionAssetExtractor.php:32]: I think you can put that to the header above, right under the copyright.

File: model/qti/CustomInteractionAsset/Extractor/BaseExtendedCustomInteractionAssetExtractor.php [hunk 6]
```diff
@@ -0,0 +1,56 @@
+<?php
+
+declare(strict_types=1);
```
Comment (wazelin) [inline model/qti/CustomInteractionAsset/Extractor/BaseExtendedCustomInteractionAssetExtractor.php:3]: This is the order of blocks as suggested by PSR-12

1. Opening <?php tag.
2. File-level docblock.
3. One or more declare statements.
4. ...

`declare()` therefore must be _under_ the file-level license block.

File: model/qti/AssetParser.php [hunk 7]
```diff
@@ -356,6 +359,17 @@ private function getXmlProperties($properties)
         return $xmls;
     }
 
+    private function extractAdvancedCustomInteractionAssets(CustomInteraction $interaction): void
+    {
+        $assetExtractor = ExtendedCustomInteractionAssetExtractorAllocator::allocateExtractor($interaction);
+        if ($assetExtractor !== null) {
```
Comment (wazelin) [inline model/qti/AssetParser.php:365]: If you decide not to implement a no-op service here, could you please consider adding an early `return` instead of a logic with an `if` block?
That improves readability, keeping the indentation not too nested.

File: model/qti/CustomInteractionAsset/ExtendedCustomInteractionAssetExtractorAllocator.php (hunk 3 above)
Comment (poyuki) [reply model/qti/CustomInteractionAsset/ExtendedCustomInteractionAssetExtractorAllocator.php:52]: I agree, in case of repeated use there will be more `if`. You pushed it through :)

File: model/qti/AssetParser.php (hunk 7 above)
Comment (poyuki) [reply model/qti/AssetParser.php:365]: i will create no-op service

File: model/qti/CustomInteractionAsset/Extractor/BaseExtendedCustomInteractionAssetExtractor.php (hunk 6 above)
Comment (poyuki) [reply model/qti/CustomInteractionAsset/Extractor/BaseExtendedCustomInteractionAssetExtractor.php:3]: agree but template here confused me a little bit 
https://oat-sa.atlassian.net/wiki/spaces/~369998401/pages/186221874/PhpStorm+live+templates

File: model/qti/CustomInteractionAsset/ExtendedCustomInteractionAssetExtractorAllocator.php (hunk 4 above)
Comment (poyuki) [reply model/qti/CustomInteractionAsset/ExtendedCustomInteractionAssetExtractorAllocator.php:42]: agree, make sense

File: model/qti/CustomInteractionAsset/ExtendedCustomInteractionAssetExtractorAllocator.php (hunk 2 above)
Comment (poyuki) [reply model/qti/CustomInteractionAsset/ExtendedCustomInteractionAssetExtractorAllocator.php:36]: LGTM, you just need to understand where this mapping will come from

File: model/qti/CustomInteractionAsset/ExtendedCustomInteractionAssetExtractorAllocator.php [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+        if (
+            $interaction instanceof ImsPortableCustomInteraction
+            || $interaction instanceof PortableCustomInteraction
+        ) {
+            if ($interaction->getTypeIdentifier() === self::TEXT_READER_INTERACTION) {
```
Comment (wazelin) [inline model/qti/CustomInteractionAsset/ExtendedCustomInteractionAssetExtractorAllocator.php:47]: Let's make `getTypeIdentifier()` part of `CustomInteraction`.

In order to keep backwards compatibility, it can simply return an empty `string` by default, while the descendants implement the real deal.

One more `if` gone :)

**Review by wazelin — CHANGES_REQUESTED** [review]
Overall, the solution seems to be clear and extendable.
I do have a number of nitpicks, though :)

File: model/qti/CustomInteractionAsset/Extractor/BaseExtendedCustomInteractionAssetExtractor.php (hunk 6 above)
Comment (wazelin) [reply model/qti/CustomInteractionAsset/Extractor/BaseExtendedCustomInteractionAssetExtractor.php:3]: Welp, you spot an issue with it :)

Comment (wazelin) [reply model/qti/CustomInteractionAsset/Extractor/BaseExtendedCustomInteractionAssetExtractor.php:3]: Here's the common doc BTW
https://oat-sa.atlassian.net/wiki/spaces/NEX/pages/87361761/How+To+Insert+Open+Assessment+Technologies+SA+Licence+Header+Comment

File: model/qti/CustomInteractionAsset/Extractor/BaseExtendedCustomInteractionAssetExtractor.php (hunk 5 above)
Comment (poyuki) [reply model/qti/CustomInteractionAsset/Extractor/BaseExtendedCustomInteractionAssetExtractor.php:32]: In some other classes this used as class dockblock

File: model/qti/CustomInteractionAsset/ExtendedCustomInteractionAssetExtractorAllocator.php (hunk 2 above)
Comment (wazelin) [reply model/qti/CustomInteractionAsset/ExtendedCustomInteractionAssetExtractorAllocator.php:36]: I suppose we can do that on the service provider level.
https://github.com/oat-sa/generis/blob/master/core/DependencyInjection/ContainerServiceProviderInterface.php

Comment (poyuki) [reply model/qti/CustomInteractionAsset/ExtendedCustomInteractionAssetExtractorAllocator.php:36]: probably in this case extractorMap should contain factories set  with some interface
```php
interface CustomInteractioAssetInstractorFactoryInterface {
   public function create(CustomInteraction $interaction): AssetExtractorInterface
}
```
and allocator in this case will be look like 
```php
public function __constuct(private array $extractorMap) {
}

public function allocateExtractor(CustomInteraction $interaction) {
    return $this->extractorMap[$interaction->getTypeIdentifier()]->create(interaction) ?? new NullAssetExtractor();
}
```
to keep flexibility in creation of some complex extractors

File: model/qti/CustomInteractionAsset/Extractor/BaseExtendedCustomInteractionAssetExtractor.php (hunk 5 above)
Comment (wazelin) [reply model/qti/CustomInteractionAsset/Extractor/BaseExtendedCustomInteractionAssetExtractor.php]: I don't think it should be that way to be honest.
We usually use a file-level docblock and that's it.

File: model/qti/CustomInteractionAsset/Extractor/Api/AssetExtractorInterface.php [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+
+/**
+ * @author Kiryl Poyu <kyril.poyu@taotesting.com>
+ */
+interface AssetExtractorInterface
+{
+    /**
+     * @return array<string>
+     */
+    public function extract(CustomInteraction $interaction): array;
```
Comment (wazelin) [inline model/qti/CustomInteractionAsset/Extractor/Api/AssetExtractorInterface.php:35]: Just a suggestion, WDYT of returning an `iterable` type here?
That will allow you to replace the implementation with a generator, instead of building the whole array of assets.

File: model/qti/CustomInteractionAsset/Extractor/TextReaderAssetExtractor.php [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+    public const CONTENT_PREFIX = 'content-';
+
+    /**
+     * @inheritDoc
+     */
+    public function extract(CustomInteraction $interaction): array
+    {
+        $extractedAssets = [];
+        $contentProperties = array_filter($interaction->getProperties(), static function ($propertyKey) {
+            return strpos($propertyKey, self::CONTENT_PREFIX) !== false;
```
Comment (wazelin) [inline model/qti/CustomInteractionAsset/Extractor/TextReaderAssetExtractor.php:45]: Considering the `'content-'` value is a prefix, WDYT of checking if the key actually starts from it, not just contains it?
```suggestion
            return strpos($propertyKey, self::CONTENT_PREFIX) === 0;
```

**Review by wazelin — CHANGES_REQUESTED** [review]

File: model/qti/CustomInteractionAsset/ExtendedCustomInteractionAssetExtractorAllocator.php (hunk 2 above)
Comment (wazelin) [reply model/qti/CustomInteractionAsset/ExtendedCustomInteractionAssetExtractorAllocator.php]: Or that map itself can be a collection of callbacks, instantiating a required extractor.

File: model/qti/CustomInteractionAsset/Extractor/BaseExtendedCustomInteractionAssetExtractor.php (hunk 5 above)
Comment (poyuki) [reply model/qti/CustomInteractionAsset/Extractor/BaseExtendedCustomInteractionAssetExtractor.php]: i think you are right, but i started do that in following this doc and other class docblock
https://oat-sa.atlassian.net/wiki/spaces/OAT/pages/4161651/PHP

File: test/unit/model/qti/CustomInteractionAsset/CustomInteractionAssetExtractorAllocatorTest.php [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+ * Copyright (c) 2021 (original work) Open Assessment Technologies SA ;
+ */
+
+namespace oat\taoQtiItem\test\unit\model\qti\CustomInteractionAsset;
+
+use oat\generis\test\TestCase;
+use oat\taoQtiItem\model\qti\CustomInteractionAsset\CustomInteractionAssetExtractorAllocator;
+use oat\taoQtiItem\model\qti\CustomInteractionAsset\Extractor\NullAssetExtractor;
+use oat\taoQtiItem\model\qti\CustomInteractionAsset\Extractor\TextReaderAssetExtractor;
+use Ramsey\Uuid\Uuid;
```
Comment (wazelin) [inline test/unit/model/qti/CustomInteractionAsset/CustomInteractionAssetExtractorAllocatorTest.php:29]: I don't think this particular extension requires that lib by default.
Please either
1. Require it as a `dev` dependency
2. Use something like `uniqid()` instead, which should be sufficient for the tests.

Comment (poyuki) [reply test/unit/model/qti/CustomInteractionAsset/CustomInteractionAssetExtractorAllocatorTest.php]: done

File: model/qti/CustomInteractionAsset/Extractor/TextReaderAssetExtractor.php (hunk 10 above)
Comment (poyuki) [reply model/qti/CustomInteractionAsset/Extractor/TextReaderAssetExtractor.php]: LGTM

File: model/qti/CustomInteractionAsset/Extractor/Api/AssetExtractorInterface.php (hunk 9 above)
Comment (poyuki) [reply model/qti/CustomInteractionAsset/Extractor/Api/AssetExtractorInterface.php]: agree with you

**Review by wazelin — CHANGES_REQUESTED** [review]
Those are not your changes, but the tests are failing on that extension. Let's fix them!

```
PHPUnit 8.5.21 by Sebastian Bergmann and contributors.


Failed asserting that two arrays are identical.
<Click to see difference>

 /Users/sergei.mikhailov/Sources/nextgen-stack/tao/taoQtiItem/test/unit/model/QtiJsonItemCompilerTest.php:169
 

TypeError : Argument 1 passed to oat\tao\model\media\MediaAsset::__construct() must implement interface oat\tao\model\media\MediaBrowser, null given, called in /Users/sergei.mikhailov/Sources/nextgen-stack/tao/taoItems/models/classes/media/ItemMediaResolver.php on line 45
 /Users/sergei.mikhailov/Sources/nextgen-stack/tao/tao/models/classes/media/MediaAsset.php:45
 /Users/sergei.mikhailov/Sources/nextgen-stack/tao/taoItems/models/classes/media/ItemMediaResolver.php:45
 /Users/sergei.mikhailov/Sources/nextgen-stack/tao/taoQtiItem/model/compile/QtiAssetCompiler/QtiItemAssetCompiler.php:109
 /Users/sergei.mikhailov/Sources/nextgen-stack/tao/taoQtiItem/model/compile/QtiAssetCompiler/QtiItemAssetCompiler.php:68
 /Users/sergei.mikhailov/Sources/nextgen-stack/tao/taoQtiItem/test/unit/model/compile/QtiItemAssetCompilerTest.php:299
```

I've just run that under PHP 7.4.

File: test/unit/model/qti/CustomInteractionAsset/CustomInteractionAssetExtractorAllocatorTest.php [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+     */
+    private $subject;
+
+    protected function setUp(): void
+    {
+        parent::setUp();
+        $this->subject = new CustomInteractionAssetExtractorAllocator([
+            TextReaderAssetExtractor::INTERACTION_IDENTIFIER => static function () {
+                return new TextReaderAssetExtractor();
+            }
```
Comment (wazelin) [inline test/unit/model/qti/CustomInteractionAsset/CustomInteractionAssetExtractorAllocatorTest.php:46]:
```suggestion
            TextReaderAssetExtractor::INTERACTION_IDENTIFIER => new TextReaderAssetExtractor(),
```

**Review by wazelin — CHANGES_REQUESTED** [review]

**Review by wazelin — APPROVED** [review]
- [x] New code is covered by tests (if applicable)
- [x] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [x] New code is respecting code style rules
- [x] New code is respecting best practices
- [x] New code is not subject to concurrency issues (if applicable)
- [x] Feature is working correctly on my local machine (if applicable)
- [x] Acceptance criteria are respected
- [x] Pull request title and description are meaningful
- [x] Pull request's target is not `master`

File: model/qti/AssetParser.php [hunk 13]
```diff
@@ -356,6 +360,16 @@ private function getXmlProperties($properties)
         return $xmls;
     }
 
+    private function extractAdvancedCustomInteractionAssets(CustomInteraction $interaction): void
+    {
+        $extractorAllocator = $this->getCustomInteractionAssetExtractorAllocator();
+        $extractor = $extractorAllocator->allocateExtractor($interaction->getTypeIdentifier());
+
+        foreach ($extractor->extract($interaction) as $asset) {
+            $this->addAsset('apip', $asset);
```
Comment (augustas) [inline model/qti/AssetParser.php:369]: Why all these extractors would add assets as `apip`? `apip` is a specific type of accessibility related stuff http://www.imsglobal.org/apip/index.html

File: composer.json [hunk 14]
```diff
@@ -61,6 +61,8 @@
     },
     "minimum-stability": "dev",
     "require": {
+        "ext-fileinfo": "*",
+        "ext-simplexml": "*",
```
Comment (augustas) [inline composer.json:65]: `ext-simplexml` is required by `qti-sdk`, [here](https://github.com/oat-sa/qti-sdk/blob/master/composer.json#L33), and this repository wouldn't work without `qti-sdk`.
Why did you need to add `ext-fileinfo` requirement?

**Review by augustas — CHANGES_REQUESTED** [review]
Great work 👍🏻  As I understand the flexibility here is to have the possibility to add asset extractors for other PCIs later, not only text reader interaction. 
Why do we need to limit the data URLs parsing only to text reader interaction? Why not make all PCI data URLs extracted? Are there custom non html tags that have to be parsed?
Could `AssetParser::loadCustomElementAssets(Element $element)` be refactored to extract the xml xpath asset parsing to the newly introduced `AssetExtractor` abstraction?

File: composer.json (hunk 14 above)
Comment (poyuki) [reply composer.json:65]: previously i used `mime_content_type` for for determining asset type. but after this behavior was simplified. And now `ext-fileinfo` can be deleted

File: model/qti/AssetParser.php (hunk 13 above)
Comment (poyuki) [reply model/qti/AssetParser.php:369]: i was trying to define type of asset by mime-type from data url, but it was increased complexity of the component by supporting mapping mime types to assets types. and i decide to use something common, cause we just need to push in assets result array and we have restriction of supported assets type. Do you see any critical problems in this case? maybe you can advise something for that usage?

**Comment by poyuki** [comment]
@augustas 
> Why do we need to limit the data URLs parsing only to text reader interaction? Why not make all PCI data URLs extracted? 

I was solving a specific problem for text reader. Probably this abstraction can be extended and improved, but for me not clear all possible aspect right now.

> Are there custom non html tags that have to be parsed?

We can cover this case by extending this component or creating additional.

> Could AssetParser::loadCustomElementAssets(Element $element) be refactored to extract the xml xpath asset parsing to the newly introduced AssetExtractor abstraction?

Probably we can do it as a tech dep task in future

**Review by yohanlaborda — CHANGES_REQUESTED** [review]

File: test/unit/model/qti/CustomInteractionAsset/Extractor/TextReaderExtendedAssetExtractorTest.php [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+     */
+    private function generateCustomInteractionWithProperties(): CustomInteraction
+    {
+        $properties = [];
+        for ($i = 0, $maxAssets = random_int($i, 10); $i < $maxAssets; $i++) {
+            $dataUrl = "data:image/jpeg;base64," . uniqid('test', true);
+            $properties[TextReaderAssetExtractor::CONTENT_PREFIX . uniqid('test', true)] = $dataUrl;
+        }
+
+        return new class($properties) extends CustomInteraction {
```
[the remark this answers is not in the document]
Comment (poyuki) [reply test/unit/model/qti/CustomInteractionAsset/Extractor/TextReaderExtendedAssetExtractorTest.php:73]: probably, but i think it's not necessary, by usage anonymous class i can emulate needed only for exact test behavior

File: model/qti/CustomInteractionAsset/Extractor/TextReaderAssetExtractor.php [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+
+use oat\taoQtiItem\model\qti\CustomInteractionAsset\Extractor\Api\AssetExtractorInterface;
+use oat\taoQtiItem\model\qti\interaction\CustomInteraction;
+
+/**
+ * @author Kiryl Poyu <kyril.poyu@taotesting.com>
+ */
+class TextReaderAssetExtractor implements AssetExtractorInterface
+{
+    public const INTERACTION_IDENTIFIER = 'textReaderInteraction';
```
[the remark this answers is not in the document]
Comment (poyuki) [reply model/qti/CustomInteractionAsset/Extractor/TextReaderAssetExtractor.php:35]: can you add more details to your comment?

File: model/qti/CustomInteractionAsset/Extractor/NullAssetExtractor.php [hunk 17]
```diff
@@ -0,0 +1,39 @@
...
+declare(strict_types=1);
+
+namespace oat\taoQtiItem\model\qti\CustomInteractionAsset\Extractor;
+
+use oat\taoQtiItem\model\qti\interaction\CustomInteraction;
+
+/**
+ * @author Kiryl Poyu <kyril.poyu@taotesting.com>
+ */
+class NullAssetExtractor implements Api\AssetExtractorInterface
```
[the remark this answers is not in the document]
Comment (poyuki) [reply model/qti/CustomInteractionAsset/Extractor/NullAssetExtractor.php:30]: good point

[the remark this answers is not in the document]
Comment (wazelin) [reply model/qti/CustomInteractionAsset/Extractor/NullAssetExtractor.php:30]: https://en.wikipedia.org/wiki/Null_object_pattern

**Comment by wazelin** [comment]
@augustas, since you're appearing in this PR (thank you for that BTW!), I'm gonna use it to send a message to you.

Please ping in Slack _today_ when you have a minute.

File: model/qti/CustomInteractionAsset/Extractor/NullAssetExtractor.php (hunk 17 above)
[the remark this answers is not in the document]
Comment (poyuki) [reply model/qti/CustomInteractionAsset/Extractor/NullAssetExtractor.php:30]: renaming the extractor to `empty` just make it semantically closer to behavior which class provide, WDYT @wazelin

[the remark this answers is not in the document]
Comment (wazelin) [reply model/qti/CustomInteractionAsset/Extractor/NullAssetExtractor.php:30]: IMO that's still a `Null` object. Let's keep it as is in order to correspond to the design pattern name.

File: model/qti/CustomInteractionAsset/Extractor/TextReaderAssetExtractor.php (hunk 16 above)
[the remark this answers is not in the document]
Comment (yohanlaborda) [reply model/qti/CustomInteractionAsset/Extractor/TextReaderAssetExtractor.php:35]: When you put a public constant, anyone can use it in another class or part of the code, that creates dependency on your code and if you change that constant it can break your code.

If you need to use it in your test, you only need to copy the value in the test, never put the reference to the constant. If someone changes the code, the test should break.

If you need the value to use it in various parts of the code, you could create an Enum (for now only available in version 8.1) or create a ValueObject that would behave like an Enum.

File: test/unit/model/qti/CustomInteractionAsset/Extractor/TextReaderExtendedAssetExtractorTest.php (hunk 15 above)
[the remark this answers is not in the document]
Comment (yohanlaborda) [reply test/unit/model/qti/CustomInteractionAsset/Extractor/TextReaderExtendedAssetExtractorTest.php:73]: When you break into class, you can put a more explicit name and reuse the class in a different test.

This is useful to clarify the code, but it is not necessary in all cases. For me it is better to create many classes with an explicit name.

File: model/qti/CustomInteractionAsset/Extractor/TextReaderAssetExtractor.php (hunk 16 above)
[the remark this answers is not in the document]
Comment (poyuki) [reply model/qti/CustomInteractionAsset/Extractor/TextReaderAssetExtractor.php:35]: I understand a concept. But i don't think it always necessary increase complexity of component in this way. it's fair for `CONTENT_PREFIX` which created in general for internal usage.

**Review by yohanlaborda — APPROVED** [review]

File: model/qti/AssetParser.php (hunk 13 above)
Comment (augustas) [reply model/qti/AssetParser.php:369]: Well, you could use a [map from elsewhere](https://github.com/thephpleague/mime-type-detection/blob/main/src/GeneratedExtensionToMimeTypeMap.php), but I agree there is no great benefit to learn the exact asset type here. Yet `apip` is a bit misleading. You used term `advanced` in the method name to differentiate theses extracted assets, maybe it could be used as a type here as well

Comment (poyuki) [reply model/qti/AssetParser.php:369]: usage of some new type there create a need to extend this map https://github.com/oat-sa/extension-tao-item/blob/master/models/classes/pack/ItemPack.php#L48

Comment (augustas) [reply model/qti/AssetParser.php:371]: Okay, but it doesn't change the fact that `apip` type here is incorrect and it's a shortcut. If no time to refactor, by returning a proper type or creating a new type, I think it would be better to add a comment why `apip` type was used here

**Review by augustas — APPROVED** [review]
- [x] New code is covered by tests (if applicable)
- [ ] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [x] New code is respecting code style rules
- [x] New code is respecting best practices
- [ ] New code is not subject to concurrency issues (if applicable)
- [x] Feature is working correctly on my local machine (if applicable)
- [x] Acceptance criteria are respected
- [x] Pull request title and description are meaningful
- [ ] Changelog is updated according to changes (if applicable)
- [ ] Documentation is updated according to changes (if applicable)
